File: python/exercise2/SSLServer.py
import socket
import ssl
import threading
from pathlib import Path
import select

LOCAL_HOST = 'localhost'
LOCAL_PORT = 8383
RESOURCE_DIRECTORY = Path(__file__).resolve().parent.parent / 'resources' / 'server'
SERVER_CERT_CHAIN = RESOURCE_DIRECTORY / 'server.intermediate.chain.pem'
SERVER_KEY = RESOURCE_DIRECTORY / 'server.key.pem'

class SSLServer:
    """
    Exercise 2 Server
    """
    def __init__(self):
        """
        Creates an SSLContext which provides parameters for any future SSL connections
        """
        context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
        context.load_cert_chain(certfile=SERVER_CERT_CHAIN, keyfile=SERVER_KEY)

        # Part 1: Server
        # Add code here to request mutual authentication from the client. You'll also need to add a reference to the
        # CA cert above
        #
        # For help check out:
        #      https://github.com/mikepound/tls-exercises/blob/master/python/README.md

        self.context = context

    def start_server(self):
        """
        Begins listening on a socket. Any connections that arrive are wrapped in an SSLSocket using the context
        created during initialisation.

        Makes use of the OS select function to perform basic non-blocking IO. Once a connection has established
        an instance of ClientHandler is created to serve the client
        """
        server_socket = socket.socket()
        server_socket.bind((LOCAL_HOST, LOCAL_PORT))
        server_socket.listen(5)
        read_list = [server_socket]

        print("Listening on port {0}...".format(LOCAL_PORT))

        while True:
            readable, writable, errored = select.select(read_list, [], [], 2)
            for s in readable:
                if s is server_socket:
                    client_socket, address = server_socket.accept()
                    try:
                        # Wrap the socket in an SSL connection (will perform a handshake)
                        conn = self.context.wrap_socket(client_socket, server_side=True)
                        ClientHandler(conn).start()
                    except ssl.SSLError as e:
                        print(e)


class ClientHandler(threading.Thread):
    """
    Thread handler leaves the main thread free to handle any other incoming connections
    """
    def __init__(self, conn):
        threading.Thread.__init__(self)
        self.conn = conn

    def run(self):
        try:
            # Read up to 1024 bytes from the client
            client_request = self.conn.recv(1024)

            # This client and server are sending encoded UTF-8 strings
            print("Received from client:", client_request.decode("UTF-8").rstrip())

            # Reply to the client's request with a small web page.
            self.conn.send("HTTP/1.0 200 OK\nContent-type: text/html\n"
                           "<html>\n"
                           "  <head>\n"
                           "    <title>Tiny Website</title>\n"
                           "  </head>\n"
                           "  <body>\n"
                           "    A very small website!\n"
                           "  </body>\n"
                           "</html>\n".encode("UTF-8"))

        except ssl.SSLError as e:
            print(e)
        except Exception as e:
            print(e)
        finally:
            self.conn.close()


def main():
    server = SSLServer()
    server.start_server()


if __name__ == '__main__':
    main()

File: java/src/main/java/com/tlsexercises/exercise2/SSLServer.java
package com.tlsexercises.exercise2;

import javax.net.ServerSocketFactory;
import javax.net.ssl.SSLServerSocket;
import javax.net.ssl.SSLServerSocketFactory;
import java.io.*;
import java.net.Socket;
import java.nio.ByteBuffer;
import java.util.Arrays;

/**
 * Exercise 2 Server
 */
public class SSLServer {

    private static final int LOCAL_PORT = 8383;
    private static final String KEYSTORE_LOCATION = SSLServer.class.getResource("/server/ServerKeyStore.jks").getPath();
    private static final String KEYSTORE_PASSWORD = "experttls";

    // Entry point
    public static void main(String argv[]) throws Exception {
        // You can also update properties as a command line parameter e.g.
        //     -Djavax.net.ssl.keyStore="keystore_location"
        System.setProperty("javax.net.ssl.keyStore", KEYSTORE_LOCATION);
        System.setProperty("javax.net.ssl.keyStorePassword", KEYSTORE_PASSWORD);

        // Part 1
        // In order to verify the client, this server requires a trust store. Use setProperty commands to
        // load a trustStore. You'll also need some static variables above as with the keystore implementation.
        //
        // For help check out:
        //      https://github.com/mikepound/tls-exercises/blob/master/java/README.md

        SSLServer server = new SSLServer();
        server.startServer();
    }

    // Start server
    public void startServer() {
        try {
            ServerSocketFactory ssf = SSLServerSocketFactory.getDefault();
            SSLServerSocket socket = (SSLServerSocket) ssf.createServerSocket(LOCAL_PORT);

            // Part 1
            // Add code here to enable client-side authentication. The server will then request the client
            // certificate during the handshake
            //
            // For help check out:
            //      https://github.com/mikepound/tls-exercises/blob/master/java/README.md

            System.out.println(String.format("Listening on port %d...", socket.getLocalPort()));

            while (true) {
                Socket client = socket.accept();
                ClientHandler handler = new ClientHandler(client);
                handler.start();
            }
        } catch (Exception e) {
            System.out.println("Exception:" + e.getMessage());
        }
    }

    /**
     * This client handler sends and receives strings via BufferedReader and BufferedWriter
     */
    class ClientHandler extends Thread {

        Socket client;
        BufferedReader input;
        BufferedWriter output;

        public ClientHandler(Socket s) { // constructor
            client = s;

            try {
                input = new BufferedReader(new InputStreamReader(client.getInputStream()));
                output = new BufferedWriter(new OutputStreamWriter(client.getOutputStream()));
            } catch (IOException e) {
                System.err.println("Error creating streams: " + e.getMessage());
            }
        }

        public void run() {
            try {
                // Obtain client message and use it
                String request = input.readLine();
                System.out.println("Received from Client: " + request);

                output.write("HTTP/1.0 200 OK\nContent-type: text/html\n" +
                        "<html>\n" +
                        "  <head>\n" +
                        "    <title>Tiny Website</title>\n" +
                        "  </head>\n" +
                        "  <body>\n" +
                        "    A very small website!\n" +
                        "  </body>\n" +
                        "</html>\n");
                output.flush();

                // Close streams and socket
                output.close();
                input.close();
                client.close();
            } catch (Exception e) {
                System.err.println("Exception: " + e.getMessage());
            }
        }
    }
}

